core/src/main/java/com/microsoft/applicationinsights/internal/statsbeat/NetworkStatsbeat.java

```diff
@@ -0,0 +1,190 @@
+            throttleCountSt.getProperties().put(CUSTOM_DIMENSIONS_INSTRUMENTATION, instrumentation);
+            telemetryClient.track(throttleCountSt);
+        }
+
+        if (local.exceptionCount.get() != 0) {
+            MetricTelemetry exceptionCountSt = createStatsbeatTelemetry(EXCEPTION_COUNT, local.exceptionCount.get());
+            exceptionCountSt.getProperties().put(CUSTOM_DIMENSIONS_INSTRUMENTATION, instrumentation);
+            telemetryClient.track(exceptionCountSt);
+        }
+    }
+
+    // TODO create a new instrumentation set automatically when its capacity reaches 64
+    public void addInstrumentation(String instrumentation) {
+        synchronized (lock) {
+            current.instrumentationList.add(instrumentation);
+        }
+    }
+
+    public Set<String> getInstrumentationList() {
+        return current.instrumentationList;
+    }
+
+    /**
+     * @return a 64-bit long that represents a list of instrumentations enabled. Each bitfield maps to an instrumentation.
+     */
+    public long getInstrumentation() {
+        return StatsbeatHelper.encodeInstrumentations(current.instrumentationList);
+    }
+
+    public void incrementRequestSuccessCount() {
+        current.requestSuccessCount.incrementAndGet();
+    }
+
+    public void incrementRequestFailureCount() {
+        current.requestFailureCount.incrementAndGet();
+    }
+
+    public void addRequestDuration(double duration) {
+        synchronized (lock) {
+            current.requestDurations.add(duration);
+        }
+    }
+
+    public void incrementRetryCount() {
+        current.retryCount.incrementAndGet();
+    }
+
+    public void incrementThrottlingCount() {
+        current.throttlingCount.incrementAndGet();
+    }
+
+    public void incrementExceptionCount() {
+        current.exceptionCount.incrementAndGet();
+    }
+
+    public long getRequestSuccessCount() {
+        return current.requestSuccessCount.get();
+    }
+
+    public long getRequestFailureCount() {
+        return current.requestFailureCount.get();
+    }
+
+    public List<Double> getRequestDurations() {
+        return current.requestDurations;
+    }
+
+    public long getRetryCount() {
+        return current.retryCount.get();
+    }
+
+    public long getThrottlingCount() {
+        return current.throttlingCount.get();
+    }
+
+    public long getExceptionCount() {
+        return current.exceptionCount.get();
+    }
+
+    public IntervalMetrics getIntervalMetrics() {
+        return current;
+    }
+
+    protected double getRequestDurationAvg(IntervalMetrics local) {
+        double sum = 0.0;
+        for (double elem : local.requestDurations) {
+            sum += elem;
+        }
+
+        if (local.requestDurations.size() != 0) {
+            return sum / local.requestDurations.size();
+        }
+
+        return  sum;
+    }
+
+    private static class IntervalMetrics {
+        private volatile Set<String> instrumentationList = new HashSet<>(64);
```

Comment:
unless reason to pre-allocate? also can be final, and should use concurrent data struct (or synchronize all access)
```suggestion
        private final Set<String> instrumentationList = Collections.newSetFromMap(new ConcurrentHashMap<>());
```